src/main/scala/mesosphere/marathon/core/externalvolume/impl/providers/DVDIProvider.scala

```diff
@@ -115,24 +115,39 @@ private[impl] object DVDIProviderValidations extends ExternalVolumeValidations {
   // task instance across the entire cluster.
   override lazy val rootGroup = new Validator[Group] {
     override def apply(g: Group): Result = {
-      val appsByVolume =
+      val appsByVolume: Map[String, Set[PathId]] =
         g.transitiveApps
           .flatMap { app => namesOfMatchingVolumes(app).map(_ -> app.id) }
           .groupBy { case (volumeName, _) => volumeName }
           .mapValues(_.map { case (volumeName, appId) => appId })
 
-      val groupViolations = g.apps.flatMap { app =>
-        val ruleViolations = namesOfMatchingVolumes(app).flatMap { name =>
-          for {
-            otherApp <- appsByVolume(name)
-            if otherApp != app.id // do not compare to self
-          } yield RuleViolation(app.id, s"Requested volume $name conflicts with a volume in app $otherApp", None)
+      val appValid: Validator[AppDefinition] = {
+        def volumeNameUnique(appId: PathId): Validator[ExternalVolume] = {
+          def conflictingApps(vol: ExternalVolume): Set[PathId] =
+            appsByVolume.getOrElse(vol.external.name, Set.empty).filter(_ != appId)
+
+          isTrue { (vol: ExternalVolume) =>
+            val conflictingAppIds = conflictingApps(vol).mkString(", ")
+            s"Volume name '${vol.external.name}' in $appId conflicts with volume(s) of same name in app(s): " +
+              s"$conflictingAppIds"
+          }{ vol => conflictingApps(vol).isEmpty }
+        }
+
+        validator[AppDefinition] { app =>
+          app.externalVolumes is every(volumeNameUnique(app.id))
```

Comment:
Then the index is wrong

Matthias Veit <notifications@github.com> schrieb am Di., 12. Apr. 2016
08:12:

> In
> src/main/scala/mesosphere/marathon/core/externalvolume/impl/providers/DVDIProvider.scala
> <https://github.com/mesosphere/marathon/pull/3737#discussion_r59325005>:
>
> > -            if otherApp != app.id // do not compare to self
> > -          } yield RuleViolation(app.id, s"Requested volume $name conflicts with a volume in app $otherApp", None)
> > +      val appValid: Validator[AppDefinition] = {
> > +        def volumeNameUnique(appId: PathId): Validator[ExternalVolume] = {
> > +          def conflictingApps(vol: ExternalVolume): Set[PathId] =
> > +            appsByVolume.getOrElse(vol.external.name, Set.empty).filter(_ != appId)
> > +
> > +          isTrue { (vol: ExternalVolume) =>
> > +            val conflictingAppIds = conflictingApps(vol).mkString(", ")
> > +            s"Volume name '${vol.external.name}' in $appId conflicts with volume(s) of same name in app(s): " +
> > +              s"$conflictingAppIds"
> > +          }{ vol => conflictingApps(vol).isEmpty }
> > +        }
> > +
> > +        validator[AppDefinition] { app =>
> > +          app.externalVolumes is every(volumeNameUnique(app.id))
>
> To have the correct path here, you can do:
>
> app.externalVolumes as "container/volumes" is every(volumeNameUnique(app.id))
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/mesosphere/marathon/pull/3737/files/8795607fb366bab5b1537cf741f15c6478c3aede#r59325005>
>